src/ai_config.js

```diff
@@ -0,0 +1,158 @@
+    aiPath: {file: "AutoIt3.exe"},
+    wrapperPath: {dir: "SciTE\\AutoIt3Wrapper\\", file: "AutoIt3Wrapper.au3"},
+    checkPath: {file: "AU3Check.exe"},
+    helpPath: {file: "AutoIt3Help.exe"},
+    infoPath: {file: "Au3Info.exe"},
+    kodaPath: {dir: "SciTE\\Koda\\", file: "FD.exe"},
+    includePaths: [{dir: ""}],
+    smartHelp: [{dir: "", file: ""}]
+  }
+};
+
+const listeners = new Map();
+let listenerId = 0, aiPath = "";
+workspace.onDidChangeConfiguration(({ affectsConfiguration }) => {
+  if (!affectsConfiguration('autoit')) return;
+
+  conf.data = workspace.getConfiguration('autoit');
+  listeners.forEach(listener => {
+    try{listener();}catch(er){console.error(er);}
+  });
+  getPaths();
+});
+
+getPaths();
+
+function addListener(listener) {
+  listeners.set(++listenerId, listener);
+  return listenerId;
+}
+
+function removeListener(id) {
+  listeners.remove(id);
+}
+
+function splitPath(path)
+{
+  path = path.trim().match(/^(.*[\\/])?([^\\/]+)?$/).map(a => a || "");
+
+  return {
+    path: path[0],
+    dir: path[1] + (path[1] === "" ? "" : "\\"),
+    file: path[2],
+    isRelative: !!(path[1] && !path[1].match(/^[a-zA-Z]:[\\/]/))
+  };
+}
+
+function fixPath (key, value, index)
+{
+  const path = splitPath(value || "");
+  const file = (conf.paths[key][index] || conf.paths[key]).file;
+  const dir = (conf.paths[key][index] || conf.paths[key]).dir;
+  if (path.file === "")
+    path.file = file || "";
+
+  if (path.dir === "" || path.isRelative)
+    path.dir = aiPath.dir + path.dir + (!path.isRelative ? (dir || "") : "");
+  
+  if (file === undefined)
+    path.file += "/";
+
+  return (path.dir + "/" + path.file).replace(/[\\/]+/g, "\\");
+}
+
+function verifyPath (val, obj, key, index)
+{
+  const showError = () =>
+  {
+    const timeout = obj.message && !obj.message.isHidden ? 1000 : 0;
+    if (timeout) {
+      obj.message.hide();
+      delete obj.message;
+    }
+
+    if (obj.prevCheck !== val)
+    {
+      const type = (conf.paths[key][index] || conf.paths[key]).file !== undefined ? "File" : "Directory";
+      setTimeout(() => obj.message = showErrorMessage(`${type} "${obj.fullPath}" not found (autoit.${key}${index === undefined ? '' : ` [${index}]`})`), timeout);
+    }
+
+    obj.prevCheck = val;
+  };
+
+  workspace.fs.stat(Uri.parse(`file:///${obj.fullPath}`)).then(data => {
+    const type = ((conf.paths[key][index] || conf.paths[key]).file !== undefined ? FileType.File : FileType.Directory) | FileType.SymbolicLink;
+    if (!(data.type & type))
+      return showError();
+
+    if (obj.message)
+    {
+      obj.message.hide();
+      delete obj.message;
+    }
+    obj.prevCheck = val;
+
+  }).catch(() =>
+  {
+    showError();
+  });
+}
+
+function getPaths()
+{
+  aiPath = splitPath(conf.data.aiPath||"");
+
+  for(let i in conf.paths)
+  {
+    const value = conf.data[i];
+    if (Array.isArray(value)) {
+      for(let j = 0; j < value.length; j++ )
+      {
+        const val = value[j].trim();
```

Comment:
@vanowm This doesn't work correctly with the 'smartHelp' entry because `value` will contain an array instead of a string.

Reply:
How exactly it "doesn't work correctly"? and what type of path do you test it with?

Reply:
Oh, I thought it was list of paths for the help files...

Reply:
After examining the code, it seems the original default value for that setting was incorrect, it contained a single array with 3 empty strings (`["", "", ""]`), however it was supposed to be an array within array (`[["", "", ""]]`). The readme shows the correct value though.
From the end user view, however, I don't think this is very intuitive or descriptive enough to understand what these values supposed to be, maybe instead of an array, we should convert it into an object:
```
[{
"prefix": "",
"path": "",
"sources: "",
}]
```
?
And also, I'm not quite understand what these `sources` are

Reply:
Yes, switching to an object would perhaps lead to better acceptance. `sources` are the actual .au3 files associated with the 3rd party help file. Here's what my current entry from settings.json looks like --

```
        "autoit.smartHelp": [
		
            [
                "_AD_",
                "C:\\Users\\danpo\\AppData\\Autoit.Advanced.Help\\AD.chm",
                "AD.au3"
            ],
            [
                "_OL_",
                "C:\\Users\\danpo\\AppData\\Autoit.Advanced.Help\\OutlookEX.chm",
                "OutlookEX.au3|OutlookEX_Base.au3"
            ],
            [
                "_TS_",
                "C:\\Users\\danpo\\AppData\\Autoit.Advanced.Help\\TaskScheduler.chm",
                "TaskScheduler.au3"
            ],
            [
                "_WD_",
                "C:\\Users\\danpo\\AppData\\Autoit.Advanced.Help\\Webdriver.chm",
                "C:\\Users\\danpo\\OneDrive\\Documents\\GitHub\\WebDriver\\wd_core.au3|C:\\Users\\danpo\\OneDrive\\Documents\\GitHub\\WebDriver\\wd_helper.au3|C:\\Users\\danpo\\OneDrive\\Documents\\GitHub\\WebDriver\\wd_cdp.au3"
            ]
			
		],
```

Reply:
I think it will be even better get rid of the array all together, since prefixes should be unique anyway. The new scheme will be:
```
"autoit.smartHelp": {
  "_PREFIX_": {
    "chmPath": "",
    "udfPath": [""]
  }
}
```

One thing confuses me though, why `.au3` files are required to open help file?

Reply:
Conversion from array based to object is automatic, user won't need do anything. Maybe it just me, but this looks more user friendly than current scheme:
```
"autoit.smartHelp": {
	"_AD_": {
		"chmPath": "C:\\Users\\danpo\\AppData\\Autoit.Advanced.Help\\AD.chm",
		"udfPath": [
			"AD.au3"
		]
	},
	"_OL_": {
		"chmPath": "C:\\Users\\danpo\\AppData\\Autoit.Advanced.Help\\OutlookEX.chm",
		"udfPath": [
			"OutlookEX.au3",
			"OutlookEX_Base.au3"
		]
	},
	"_TS_": {
		"chmPath": "C:\\Users\\danpo\\AppData\\Autoit.Advanced.Help\\TaskScheduler.chm",
		"udfPath": [
			"TaskScheduler.au3"
		]
	},
	"_WD_": {
		"chmPath": "C:\\Users\\danpo\\AppData\\Autoit.Advanced.Help\\Webdriver.chm",
		"udfPath": [
			"C:\\Users\\danpo\\OneDrive\\Documents\\GitHub\\WebDriver\\wd_core.au3",
			"C:\\Users\\danpo\\OneDrive\\Documents\\GitHub\\WebDriver\\wd_helper.au3",
			"C:\\Users\\danpo\\OneDrive\\Documents\\GitHub\\WebDriver\\wd_cdp.au3"
		]
	}
},
```

About sources, that's exactly what confuses me, why does it need to search sources at all? I mean, shouldn't it search the help file instead, to make sure that function exists there?
For example, `_WD_JsonActionPointer` doesn't exist in help file, but it exists in sources, so it opens help file with error page.